accepted/targeting-packs/targeting-packs.md

```diff
@@ -0,0 +1,149 @@
+
+This proposal is about enabling acquisition of the .NET Framework targeting
+packs via NuGet so that compiling code for a specific version of the .NET
+Framework doesn't require a Windows machine with a machine-wide installation of
+the corresponding targeting pack.
+
+Since .NET Framework 2.0, the .NET platform supported targeting different
+versions of the platform regardless of the .NET Framework version that is
+installed on the developer's machine. This allows developers to target a lower
+version than what they use themselves. This is especially useful for in-place
+updates where the older version of the framework cannot be installed side-by-
+side.
+
+That mechanism is called *targeting packs*, which is simply a set of assemblies
+that describe the public APIs of a particular version of the .NET Framework.
+Since those assemblies are only used by the compiler, they don't need to contain
+any code, which we call [reference assemblies](https://www.youtube.com/watch?v=EBpY1UMHDY8).
+
+Unfortunately, the targeting packs aren't available by themselves; they are
+redistributed as part of the .NET Framework Developer Packs which bundle both
+the runtime as well as the targeting pack. This design ensures that developers
+can never be in a state where they have a targeting pack that is higher than the
+.NET Framework version they have installed. That design made sense coming from a
+Visual Studio-first perspective but doesn't serve us well today:
+
+* **It's not cross-platform friendly**. The .NET Framework Developer Packs are
+  only available for Windows which makes the targeting packs indirectly Windows-
+  only too.
+
+* **It's not CI friendly**. Requiring a machine-wide install isn't friendly for
+  cloud-hosted CI machines as developers typically lack permissions to install
+  software.
+
+This document proposes to offer the targeting packs via NuGet that they can be
+acquired in a non-impactful way. It builds on the following existing specs:
+
+* [Phillip Carter](https://github.com/cartermp): [Add Target Packs on NuGet](https://github.com/dotnet/core/pull/64)
+* [Daniel Plaisted](https://github.com/dsplaisted): [Reference Assembly NuGet packages](https://gist.github.com/dsplaisted/83d67bbcff9ec1d0aff1bea1bf4ad79a)
+* [Prototype](https://github.com/dsplaisted/ReferenceAssemblyPackages)
+
+## Scenarios and User Experience
+
+* Build a project which is multi-targeted to .NET Core or .NET Standard as well
+  as .NET Framework on Mac OS or Linux using the .NET CLI
+
+* Build a project targeting Mono (which uses the same Target Framework
+  Identifier as .NET Framework) using the .NET CLI
+
+## Requirements
+
+### Goals
+
+* Doesn't require running an installer
+* Doesn't require Windows
+* Can be used in SDK-style multi-targeting projects
+
+### Non-Goals
+
+* Supporting NuGet-based acquisition of the .NET Framework runtime
+* Supporting NuGet-based acquisition from non-SDK-style projects
+* Support pre-.NET Framework 4.5 targeting packs
```

Comment:
😭 Libraries like NUnit could really use this. This is what is preventing the dotnet CLI from being able to build net40 and earlier. What's the cost of two or three more targeting packs?

Reply:
Please provide all PCL and .NET Framework 2+ packages. They are already shipped out and not going to get any updates anyway, so hopefully no additional cost than just package once and push to nuget..

Reply:
I think we should provide the .NET 4.0 targeting pack, and possibly other 4.0.x versions in reference assemblies.  We already have targeting packs for these, so I don't think there are any obstacles to this.

For .NET 3.5.1 and lower, we didn't ship targeting packs in the same way.  PCL reference assemblies are also a bit different.  I would propose that initially we not support them, in order to get basic .NET Framework support out more quickly.

Reply:
@dsplaisted If net35 is merely deprioritized without closing the door on them, I can cheer up. 😃 (net20 would also make some folks happier, and by extension, me.)

Reply:
This would also fix cases, like in Roslyn, where custom packages are created

Reply:
I have talked to @jaredpar and it seems you guys don't need pre-.NET Framework 4.5 support. Is that not true?

Reply:
@jnm2 

> What's the cost of two or three more targeting packs?

It's mostly consistency in our offerings; we don't want to breath more life into framework that are that old. That being said, we can always ship more targeting packs in the future if that seems necessary. Based on adoption I don't see enough evidence that this is necessary though.

Reply:
@dsplaisted 

> I think we should provide the .NET 4.0 targeting pack, and possibly other 4.0.x versions in reference assemblies. We already have targeting packs for these, so I don't think there are any obstacles to this.

That's a fair point. How about this: I leave the wording in the spec as-is; if we can easily build the 4.0 set, great, we'll ship it. If not or we run out of time, we don't. Usage wise, I don't see compelling reasons to spend a great deal of time on 4.0. And 4.5 just makes sense from a .NET Standard support matrix stand point.

Reply:
@terrajobst 

> It's mostly consistency in our offerings; we don't want to breath more life into frameworks that are that old.

So long as VSTest ships a net35-compiled test execution engine, that puts pressure on NUnit to ship a net35 assembly so that we aren't the ones preventing people from testing on CLR v2 if they need to. A net35 targeting pack would be preferable to what we're doing now, using Mono's edition of net35.

Reply:
Not sure I buy that: xUnit only supports .NET Standard 1.1. And the number of 3.5 installs is quite low. I think I'm fine with 4.0 but I'm definitely opposed to go lower, especially because the whole 2.0, 3.0, 3.5 TPs are much more convoluted.